osquery/tables/system/darwin/disk_encryption.mm

```diff
@@ -130,33 +136,118 @@ void genFDEStatusForBSDName(const std::string& bsd_name,
   r["name"] = kDeviceNamePrefix + bsd_name;
   r["uuid"] = uuid;
 
-  auto encrypted = getIOKitProperty(properties, kCoreStorageIsEncryptedKey_);
-  if (encrypted.empty()) {
-    r["encrypted"] = "0";
+  if (isAPFS) {
+
+
+    auto bundle_url = CFURLCreateWithFileSystemPath(
+      kCFAllocatorDefault,
+      CFSTR("/System/Library/PrivateFrameworks/DiskManagement.framework"),
+      kCFURLPOSIXPathStyle,
+      true);
+    if (bundle_url == nullptr) {
+      LOG(ERROR) << "parsing bundle URL";
+      return;
+    }
+
+    auto bundle = CFBundleCreate(kCFAllocatorDefault, bundle_url);
+    CFRelease(bundle_url);
+    if (bundle == nullptr) {
+      LOG(ERROR) << "opening bundle";
+      return;
+   }
+
+    CFBundleLoadExecutable(bundle);
+
+    DASessionRef session = DASessionCreate(kCFAllocatorDefault);
+
+    // DMManager *man = [DMManager sharedManager]aa
+    id cls = NSClassFromString(@"DMManager");
+    SEL sls = NSSelectorFromString(@"sharedManager");
+    id man = [cls performSelector:sls];
+
+    // Seems to be unnecessary?
+    // [man setDefaultDASession:session];
+    // TODO try/catch
+    //sls = NSSelectorFromString(@"setDefaultDASession:");
+    //[man performSelector:sls withObject:(__bridge id)session];
+
+    // DMAPFS * apfs = [[DMAPFS alloc] initWithManager:man];
+    cls = NSClassFromString(@"DMAPFS");
+    sls = NSSelectorFromString(@"alloc");
+    id apfs = [cls performSelector:sls];
+    sls = NSSelectorFromString(@"initWithManager:");
+    [apfs performSelector:sls withObject:man];
+
+    DADiskRef targetVol = DADiskCreateFromBSDName(NULL, session, r["name"].c_str());
+
+    //err = [apfs isEncryptedVolume:targetVol encrypted:&isEncrypted];
```

Comment:
My research indicated that I had to use `NSInvocation` rather than `performSelector` when passing C types to the receiver. This also doesn't generate warnings from the compiler about possible memory leaks, but I don't know that it is actually any safer.

Reply:
This is noted in the [documentation for performSelector](https://developer.apple.com/documentation/objectivec/1418956-nsobject/1418764-performselector):
> This method is the same as performSelector: except that you can supply an argument for aSelector. aSelector should identify a method that takes a single argument of type id. For methods with other argument types and return values, use NSInvocation.

Reply:
There may also be a possibility of wrapping with NSValue and using performSelector as described in (https://stackoverflow.com/a/1185566/491710) and other StackOverflow discussions.